protobuf-converter/src/main/java/io/confluent/connect/protobuf/ProtobufData.java

```diff
@@ -0,0 +1,952 @@
+
+  public static final String PROTOBUF_TYPE_ENUM = NAMESPACE + ".Enum";
+  public static final String PROTOBUF_TYPE_ENUM_PREFIX = PROTOBUF_TYPE_ENUM + ".";
+  public static final String PROTOBUF_TYPE_UNION = NAMESPACE + ".Union";
+  public static final String PROTOBUF_TYPE_UNION_PREFIX = PROTOBUF_TYPE_UNION + ".";
+  public static final String PROTOBUF_TYPE_TAG = NAMESPACE + ".Tag";
+
+  public static final String GOOGLE_PROTOBUF_PACKAGE = "google.protobuf";
+  public static final String GOOGLE_PROTOBUF_TIMESTAMP_NAME = "Timestamp";
+  public static final String GOOGLE_PROTOBUF_TIMESTAMP_FULL_NAME = GOOGLE_PROTOBUF_PACKAGE
+      + "."
+      + GOOGLE_PROTOBUF_TIMESTAMP_NAME;
+  public static final String GOOGLE_PROTOBUF_TIMESTAMP_LOCATION = "google/protobuf/timestamp.proto";
+
+  private int defaultSchemaNameIndex = 0;
+
+  private final Cache<Schema, ProtobufSchema> fromConnectSchemaCache;
+  private final Cache<ProtobufSchema, Schema> toConnectSchemaCache;
+
+  public ProtobufData() {
+    this(new ProtobufDataConfig.Builder().with(
+        SCHEMAS_CACHE_SIZE_CONFIG,
+        SCHEMAS_CACHE_SIZE_DEFAULT
+    ).build());
+  }
+
+  public ProtobufData(int cacheSize) {
+    this(new ProtobufDataConfig.Builder().with(SCHEMAS_CACHE_SIZE_CONFIG, cacheSize).build());
+  }
+
+  public ProtobufData(ProtobufDataConfig protobufDataConfig) {
+    fromConnectSchemaCache =
+        new SynchronizedCache<>(new LRUCache<>(protobufDataConfig.getSchemasCacheSize()));
+    toConnectSchemaCache =
+        new SynchronizedCache<>(new LRUCache<>(protobufDataConfig.getSchemasCacheSize()));
+  }
+
+  /**
+   * Convert this object, in Connect data format, into an Protobuf object.
+   */
+  public ProtobufSchemaAndValue fromConnectData(Schema schema, Object value) {
+    ProtobufSchema protobufSchema = fromConnectSchema(schema);
+    // Reset the default schema name index.
+    // Note: the assumption is that default names will be generated in an order in both the
+    // schema and data
+    // such that the names correspond to each other in the schema and data.
+    // TODO try to break this assumption
+    defaultSchemaNameIndex = 0;
+    return new ProtobufSchemaAndValue(
+        protobufSchema,
+        fromConnectData(schema, "", value, protobufSchema)
+    );
+  }
+
+  // Visible for testing
+  protected ProtobufSchemaAndValue fromConnectData(SchemaAndValue schemaAndValue) {
+    return fromConnectData(schemaAndValue.schema(), schemaAndValue.value());
+  }
+
+  private Object fromConnectData(
+      Schema schema,
+      String scope,
+      Object value,
+      ProtobufSchema protobufSchema
+  ) {
+    if (value == null) {
+      // Ignore missing values
+      return null;
+    }
+
+    final Schema.Type schemaType = schema.type();
+    try {
+      switch (schemaType) {
+        case INT8:
+        case INT16:
+        case INT32: {
+          final int intValue = ((Number) value).intValue(); // Check for correct type
```

Comment:
Do these type casts need to be checked before being executed ?

Reply:
The casts are executed in a try-block that catches ClassCastException and throws a DataException.